File: Arrays/container-with-most-water/container-with-most-water.py
#  Project : SDE Placement Questions
#  Filename : container-with-most-water.py
#  Author : thameem
#  Current modification time : Sat, 13 May 2023 at 11:56 pm India Standard Time
#  Last modified time : Sat, 13 May 2023 at 11:56 pm India Standard Time

# Link: https://leetcode.com/problems/container-with-most-water/submissions/


def get_max_area(height: list[int]) -> int:
    no_of_lines = len(height)
    left, right = 0, no_of_lines - 1
    max_area = 0
    while left < right:

        current_area = min(height[left], height[right]) * (right - left)
        max_area = max(max_area, current_area)

        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area

File: Arrays/container-with-most-water/container-with-most-water.cpp

#include <vector>
#include <iostream>

using namespace std;

int maxArea(vector<int> &height)
{
    int no_of_lines = height.size();
    int max_area = 0;
    int left = 0;
    int right = no_of_lines - 1;
    while (left < right)
    {
        int current_area = min(height[left], height[right]) * (right - left);
        max_area = max(current_area, max_area);

        if (height[left] < height[right])
        {
            left++;
        }
        else
        {
            right--;
        }
    }

    return max_area;
}

int main()
{
    vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    
    cout<<maxArea(height)<<"\n";

    return 0;
}